### DISCUSSION TITLE
How to set Kulala keys for .http files only

### DISCUSSION THREAD

**Discussion by lucasrabiec** [LazyVim/LazyVim#4573] [Q&A] [answered]
I don't know if I am missing something but keys are available globally (also listed in which-key). I want them to work only in .http files. Here is my config:
```lua
return {
  "mistweaverco/kulala.nvim",
  ft = "http",
  opts = {
    default_view = "headers_body",
    contenttypes = {
      ["application/problem+json"] = {
        formatter = { "jq", "." },
        ft = "json",
      },
    },
  },
  keys = {
    { "<leader>R", "", desc = "+Rest", ft = "http" },
    { "<leader>Rs", "<cmd>lua require('kulala').run()<cr>", desc = "Send the request", ft = "http" },
    { "<leader>Rt", "<cmd>lua require('kulala').toggle_view()<cr>", desc = "Toggle headers/body", ft = "http" },
    { "<leader>Rp", "<cmd>lua require('kulala').jump_prev()<cr>", desc = "Jump to previous request", ft = "http" },
    { "<leader>Rn", "<cmd>lua require('kulala').jump_next()<cr>", desc = "Jump to next request", ft = "http" },
  },
}
``` 
I thought setting `ft` in each key would work, but I was wrong. Any ideas?

**Comment by dpetka2001** [comment, marked as answer]
You should do 
```lua
keys = function()
  return {
    -- your changed keys here
  }
end,
```
Why? In very simple terms, this is because of how `lazy.nvim` parses the `keys` handler of a specific plugin. For every different keymap, `lazy.nvim` creates a `id` that differentiates the mapping from others. Usually, this `id` is resolved from just the `lhs`. But when a plugin also has a `ft` mapping, that `ft` gets concatenated with the `lhs` into the `id` specific for the corresponding mapping. Relevant code can be found [here](https://github.com/folke/lazy.nvim/blob/1159bdccd8910a0fd0914b24d6c3d186689023d9/lua/lazy/core/handler/keys.lua#L32-L53) for the parsing phase and later on you can see in the `resolve` and `add/del` functions that it uses `keys.id` to set the mappings.

Example
```lua
{ "<leader>Rs", "<cmd>lua require('kulala').run()<cr>", desc = "Send the request" },
{ "<leader>Rs", "<cmd>lua require('kulala').run()<cr>", desc = "Send the request", ft = "http" },
```
The `id` for the first mapping would be `" Rs"`, while for the second mapping the `id` would be `" Rs (http)"`. Because of this, you can have different mappings with the same `lhs`, but one will be global and the other will be filetype specific.

The way to work around this, would normally be to disable the global mapping and then specify the filetype mapping. So, you would do 
```lua
keys = {
  { "<leader>Rs", false },
  { "<leader>Rs", "<cmd>lua require('kulala').run()<cr>", desc = "Send the request", ft = "http" },
}
```
But you would have to do this for every mapping, thus duplicating the entries. The simpler approach would be (since you want to change all the mappings available) to use a `keys` function, like the one I mentioned first, which just returns the mappings you want, thus overwriting completely the default ones making them obsolete.

TLDR: Sorry for the extended answer. Bottom line is that you can have 2 different mappings with the same `lhs` one global and one filetype specific. To work around this you have to use a `keys` function to completely overwrite the default keys and only return your custom changes.

**Reply by lucasrabiec** [reply]
Thank you for such a comprehensive answer. Now this makes sense and it works. I think `ft` should be set by default for each key.

**Reply by dpetka2001** [reply]
You can make a PR to change them and the maintainer will decide. I believe there's also another PR that wants to add more keys, so maybe you could express your opinion in that PR to make the mappings filetype specific to see if the OP is willing to accommodate the changes you propose. 

Just note that the maintainer is on vacation until the end of month as stated in the corresponding issue.

**Reply by lucasrabiec** [reply]
Yea I know about vacation. I also added a PR. I appreciate your help.
